ledger/src/leader_schedule.rs

```diff
@@ -22,35 +23,60 @@ pub struct LeaderSchedule {
 
 impl LeaderSchedule {
     // Note: passing in zero stakers will cause a panic.
-    pub fn new(ids_and_stakes: &[(Pubkey, u64)], seed: [u8; 32], len: u64, repeat: u64) -> Self {
-        let (ids, stakes): (Vec<_>, Vec<_>) = ids_and_stakes.iter().cloned().unzip();
-        let rng = &mut ChaChaRng::from_seed(seed);
+    pub fn new_keyed_by_validator_identity(
+        epoch_staked_nodes: &HashMap<Pubkey, u64>,
+        epoch: Epoch,
+        len: u64,
+        repeat: u64,
+    ) -> Self {
+        let keyed_stakes: Vec<_> = epoch_staked_nodes
+            .iter()
+            .map(|(pubkey, stake)| (pubkey, *stake))
+            .collect();
+        let slot_leaders = Self::stake_weighted_slot_leaders(keyed_stakes, epoch, len, repeat);
+        Self::new_from_schedule(slot_leaders)
+    }
+
+    // Note: passing in zero stakers will cause a panic.
+    fn stake_weighted_slot_leaders(
+        mut keyed_stakes: Vec<(&Pubkey, u64)>,
```

Comment:
nit: you could avoid the intermediate collection here by taking in an iter (flipped) `Iterator<Item = (u64, &Pubkey)>` and using the iter variants in `sort_stakes`:
```
stakes.sorted_unstable().dedup().reverse()
```

Reply:
IterTools often allocates internally in its helper methods FYI:

```rust
fn sorted_unstable(..) {
        let mut v = Vec::from_iter(self);
        v.sort_unstable();
        v.into_iter()
}
```